File: generated_code/cpp/solution_98_3.cpp
#include <iostream>
#include <limits>

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    /**
     * Algorithm: Recursive Range Validation
     * Time Complexity: O(N) where N is the number of nodes in the tree, as we visit each node once.
     * Space Complexity: O(H) where H is the height of the tree, representing the recursion stack depth.
     * 
     * To be a valid BST, every node must satisfy a specific range (min, max).
     * Using long long handles edge cases where node values are INT_MIN or INT_MAX.
     */
    bool isValidBST(TreeNode* root) {
        return validate(root, std::numeric_limits<long long>::min(), std::numeric_limits<long long>::max());
    }

private:
    bool validate(TreeNode* node, long long min_val, long long max_val) {
        // Base case: An empty tree/leaf child is a valid BST
        if (!node) {
            return true;
        }

        // The current node's value must be strictly within the range (min_val, max_val)
        if (node->val <= min_val || node->val >= max_val) {
            return false;
        }

        // Recursively check subtrees:
        // Left child must be less than current node value (updates max_val)
        // Right child must be greater than current node value (updates min_val)
        return validate(node->left, min_val, node->val) && 
               validate(node->right, node->val, max_val);
    }
};

File: generated_code/python/solution_98_3.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        """
        Determines if a binary tree is a valid Binary Search Tree (BST).
        
        Algorithm: Recursive Depth-First Search (DFS) with a valid range.
        - For each node, its value must be strictly greater than a minimum bound (`low`)
          and strictly less than a maximum bound (`high`).
        - When moving left, the upper bound updates to the current node's value.
        - When moving right, the lower bound updates to the current node's value.
        
        Complexity:
        - Time Complexity: O(N) since we visit every node exactly once.
        - Space Complexity: O(N) in the worst case (skewed tree) due to the recursion stack,
                            or O(log N) for a balanced tree.
        """
        def validate(node: Optional[TreeNode], low: float, high: float) -> bool:
            # An empty tree/node is valid
            if not node:
                return True
            
            # The current node's value must strictly stay within the low and high boundaries
            if not (low < node.val < high):
                return False
            
            # Recursively validate the left and right subtrees with updated bounds
            return validate(node.left, low, node.val) and validate(node.right, node.val, high)
        
        # Initialize the boundaries with negative and positive infinity to handle all integer limits
        return validate(root, float('-inf'), float('inf'))